beps/0012-metrics-service/README.md

```diff
@@ -0,0 +1,351 @@
+    serviceVersion?: string;
+    environment?: string;
+  };
+
+  collection?: {
+    exportIntervalMillis?: number;
+    // ...
+  };
+
+  exporters: Array<{
+    type: 'prometheus' | 'otlp' | 'console' | '...';
+    config?: Record<string, any>;
+  }>;
+
+  autoInstrumentation: {
+    enabled: boolean;
+    include?: string[];
+    exclude?: string[];
+  };
+}
+```
+
+```yaml
+backend:
+  metrics:
+    enabled: true
+
+    resource:
+      serviceName: backstage
+      serviceVersion: 0.0.1
+      environment: production
+
+    # Collection settings
+    collection:
+      exportIntervalMillis: 15000
+
+    exporters:
+      - type: prometheus
+        config:
+          port: 9464
+      # ...
+      - type: console
+
+    autoInstrumentation:
+      enabled: true
+      exclude: ['express']
+```
+
+### Interface
+
+Provide a wrapper around OpenTelemetry's API while re-exporting the types from the `@opentelemetry/api` package. This introduces concepts already familiar to both the Backstage community and those familiar with OpenTelemetry.
+
+```ts
+interface MetricsService {
+  // Synchronous instrumentation
+  createCounter(name: string, options?: MetricOptions): Counter;
+  createUpDownCounter(name: string, options?: MetricOptions): UpDownCounter;
+  createHistogram(name: string, options?: MetricOptions): Histogram;
+  createGauge(name: string, options?: MetricOptions): Gauge;
+
+  // Asynchronous instrumentation
+  createObservableCounter(
+    name: string,
+    options?: MetricOptions,
+  ): ObservableCounter;
+  createObservableUpDownCounter(
+    name: string,
+    options?: MetricOptions,
+  ): ObservableUpDownCounter;
+  createObservableGauge(name: string, options?: MetricOptions): ObservableGauge;
+
+  // Future - add additional convenience methods as we learn more about the needs of the framework
+}
+```
+
+#### Root Metrics Service
+
+The `RootMetricsService` is responsible for initializing the OpenTelemetry SDK and creating plugin-scoped metrics services. If the end user wants to initialize their own SDK, they are responsible for initializing the OpenTelemetry SDK with their own configuration. The `RootMetricsService` is responsible for providing metrics to other root services and creating plugin-scoped metrics services.
+
+```ts
+interface RootMetricsService {
+  forPlugin(pluginId: string): MetricsService;
+}
+
+export const rootMetricsServiceFactory = createServiceFactory({
+  // depends on as little as possible so that it can be initialized as early as possible.
+  service: rootMetricsServiceRef,
+  deps: {
+    rootConfig: coreServices.rootConfig,
+  },
+  factory: ({ rootConfig }) => {
+    return DefaultRootMetricsService.fromConfig(rootConfig);
+  },
+});
+```
+
+```ts
+class DefaultRootMetricsService implements RootMetricsService {
+  private sdk: NodeSDK;
+
+  static fromConfig(config: Config): RootMetricsService {
+    const metricsConfig = config.getOptionalConfig('backend.metrics');
+
+    const sdk = new NodeSDK({
```

Comment:
Was it really sufficient to have this here in "normal" code? I'd expect this code to be run too late in the startup sequence. The current recommended approach is to have a bootstrap file that's run via `--include` for this, to ensure that it runs even before any regular-code `import` could ruin any global state. If we don't do that, the sdk just doesn't work fully and different parts of the system end up with different competing meter instances. What's your take on that?

Reply:
The end goal is to be config driven which would require starting the SDK in some fashion similar to the example provided. If end users wanted different, they would provide their own override where they handle the SDK themselves. 

I say we push off trying to anwer this now since we are going to revise the BEP to exclude any changes to current guidance on starting the SDK via `--require` flag.